test/TestImportIncludeRole.py

```diff
@@ -55,56 +52,54 @@
-
-    def test_include_role(self):
-        fh = self._get_play_file(PLAY_INCLUDE_ROLE)
-        runner = Runner(self.rules, fh.name, [], [], [])
-        results = runner.run()
-        assert 'only when shell functionality is required' in str(results)
-        assert 'All tasks should be named' in str(results)
-
-    def test_include_role_inline_args(self):
-        fh = self._get_play_file(PLAY_INCLUDE_ROLE_INLINE)
-        runner = Runner(self.rules, fh.name, [], [], [])
-        results = runner.run()
-        assert 'only when shell functionality is required' in str(results)
-        assert 'All tasks should be named' in str(results)
-
-    def test_invalid_import_role(self):
-        fh = self._get_play_file(PLAY_IMPORT_ROLE_INCOMPLETE)
-        runner = Runner(self.rules, fh.name, [], [], [])
-        with pytest.raises(RuntimeError, match="Failed to find required 'name' key in import_role"):
-            runner.run()
+def _setup_play(base_dir, playbook_text):
+    role_tasks_dir = base_dir / 'test-role' / 'tasks'
+    role_tasks_dir.mkdir(parents=True)
+    (role_tasks_dir / 'main.yml').write_text(ROLE_TASKS_MAIN)
+    (role_tasks_dir / 'world.yml').write_text(ROLE_TASKS_WORLD)
+    play_path = base_dir / 'playbook.yml'
+    play_path.write_text(playbook_text)
+    return str(play_path)
+
+
+@pytest.mark.parametrize(('playbook_filename', 'messages'), (
+    pytest.param(PLAY_IMPORT_ROLE,
+                 ['only when shell functionality is required'],
+                 id='IMPORT_ROLE',
+                 ),
+    pytest.param(PLAY_IMPORT_ROLE_INLINE,
+                 ['only when shell functionality is require'],
+                 id='IMPORT_ROLE_INLINE',
+                 ),
+    pytest.param(PLAY_INCLUDE_ROLE,
+                 ['only when shell functionality is require',
+                  'All tasks should be named'],
+                 id='INCLUDE_ROLE',
+                 ),
+    pytest.param(PLAY_INCLUDE_ROLE_INLINE,
+                 ['only when shell functionality is require',
+                  'All tasks should be named'],
+                 id='INCLUDE_ROLE_INLINE',
+                 ),
+))
+def test_import_role2(tmp_path, default_rules_collection, playbook_filename, messages):
+    playbook = _setup_play(tmp_path, playbook_filename)
+    runner = Runner(default_rules_collection, playbook, [], [], [])
+
+    results = runner.run()
+
+    for message in messages:
+        assert message in str(results)
+
+
+@pytest.mark.parametrize(('playbook_filename', 'error'), (
+    pytest.param(PLAY_IMPORT_ROLE_INCOMPLETE,
+                 "Failed to find required 'name' key in import_role",
+                 id='IMPORT_ROLE_INCOMPLETE',
+                 ),
+))
+def test_invalid_import_role(tmp_path, default_rules_collection, playbook_filename, error):
+    playbook = _setup_play(tmp_path, playbook_filename)
+    runner = Runner(default_rules_collection, playbook, [], [], [])
```

Comment:
How about a more classical approach?
```suggestion
@pytest.fixture
def playbook_path(request, tmp_path):
    playbook_text = request.param
    role_tasks_dir = tmp_path / 'test-role' / 'tasks'
    role_tasks_dir.mkdir(parents=True)
    (role_tasks_dir / 'main.yml').write_text(ROLE_TASKS_MAIN)
    (role_tasks_dir / 'world.yml').write_text(ROLE_TASKS_WORLD)
    play_path = tmp_path / 'playbook.yml'
    play_path.write_text(playbook_text)
    return str(play_path)


@pytest.mark.parametrize(('playbook_path', 'messages'), (
    pytest.param(PLAY_IMPORT_ROLE,
                 ['only when shell functionality is required'],
                 id='IMPORT_ROLE',
                 ),
    pytest.param(PLAY_IMPORT_ROLE_INLINE,
                 ['only when shell functionality is require'],
                 id='IMPORT_ROLE_INLINE',
                 ),
    pytest.param(PLAY_INCLUDE_ROLE,
                 ['only when shell functionality is require',
                  'All tasks should be named'],
                 id='INCLUDE_ROLE',
                 ),
    pytest.param(PLAY_INCLUDE_ROLE_INLINE,
                 ['only when shell functionality is require',
                  'All tasks should be named'],
                 id='INCLUDE_ROLE_INLINE',
                 ),
), indirect=('playbook_path', ))
def test_import_role2(default_rules_collection, playbook_path, messages):
    runner = Runner(default_rules_collection, playbook_path, [], [], [])

    results = runner.run()

    for message in messages:
        assert message in str(results)


@pytest.mark.parametrize(('playbook_path', 'error'), (
    pytest.param(PLAY_IMPORT_ROLE_INCOMPLETE,
                 "Failed to find required 'name' key in import_role",
                 id='IMPORT_ROLE_INCOMPLETE',
                 ),
), indirect=('playbook_path', ))
def test_invalid_import_role(default_rules_collection, playbook_path, error):
    runner = Runner(default_rules_collection, playbook_path, [], [], [])
```

Reply:
Didn't know about  `indirect`, neat. Though I am not sure that at first glance, it makes things much easier to follow. It's a lot of pytest _"shazam !"_, but when you know the trick, there is no more magic...

Reply:
> but when you know the trick, there is no more magic...

Right: “Any sufficiently advanced technology is indistinguishable from magic.” --Arthur C. Clarke

I think this is true for anything new to us. I knew about this feature for years before I actually stopped by to read the docs on how do I apply it to my fixtures.

In my opinion, it's important to document such things as a usage guide. Mostly, it'd be useful to have such fixtures in `conftest.py`, reused everywhere along with some guidelines. It'd confuse newbies, no doubt. And I am often having a hard time finding that golden mean between "too complicated" and "too much copy-paste".